mypy/stubgen.py

```diff
@@ -892,86 +843,217 @@ def get_qualified_name(o: Expression) -> str:
-        raise SystemExit('recursive stub generation without importing is not currently supported')
-    sigs = {}  # type: Any
-    class_sigs = {}  # type: Any
-    if options.doc_dir:
-        all_sigs = []  # type: Any
-        all_class_sigs = []  # type: Any
-        for path in glob.glob('%s/*.rst' % options.doc_dir):
-            with open(path) as f:
-                func_sigs, class_sigs = parse_all_signatures(f.readlines())
-            all_sigs += func_sigs
-            all_class_sigs += class_sigs
-        sigs = dict(find_unique_signatures(all_sigs))
-        class_sigs = dict(find_unique_signatures(all_class_sigs))
-    for module in (options.modules if not options.recursive else walk_packages(options.modules)):
-        try:
-            generate_stub_for_module(module,
-                                     output_dir=options.output_dir,
-                                     add_header=True,
-                                     sigs=sigs,
-                                     class_sigs=class_sigs,
-                                     pyversion=options.pyversion,
-                                     no_import=options.no_import,
-                                     search_path=options.search_path,
-                                     interpreter=options.interpreter,
-                                     include_private=options.include_private)
-        except Exception as e:
-            if not options.ignore_errors:
-                raise e
-            else:
-                print("Stub generation failed for", module, file=sys.stderr)
+    This is used if user passes --no-import, and will not find C modules.
+    Exit if some of the modules or packages can't be found.
+    """
+    result = []  # type: List[StubSource]
+    search_paths = SearchPaths(('.',) + tuple(search_path), (), (), ())
+    cache = FindModuleCache(search_paths)
+    for module in modules:
+        module_path = cache.find_module(module)
+        if not module_path:
+            fail_missing(module)
+        result.append(StubSource(module, module_path))
+    for package in packages:
+        p_result = cache.find_modules_recursive(package)
+        if not p_result:
+            fail_missing(package)
+        sources = [StubSource(m.module, m.path) for m in p_result]
+        result.extend(sources)
+    return result
+
+
+def mypy_options(stubgen_options: Options) -> MypyOptions:
+    """Generate mypy options using the flag passed by user."""
+    options = MypyOptions()
+    options.follow_imports = 'skip'
+    options.incremental = False
+    options.ignore_errors = True
+    options.semantic_analysis_only = True
+    options.python_version = stubgen_options.pyversion
+    return options
+
+
+def generate_asts_for_modules(py_modules: List[StubSource],
+                              parse_only: bool, mypy_options: MypyOptions) -> None:
+    """Use mypy to parse (and optionally analyze) source files."""
+    if parse_only:
+        for mod in py_modules:
```

Comment:
Optional idea for a refactoring: Move this block to a separate function (and maybe move the alternative approach to another function).